### ISSUE TITLE
Not possible to test for being a simulation model

### ISSUE THREAD

**Issue by henrikt-ma** [modelica/ModelicaSpecification#3436]
Breaking out this concern from https://github.com/modelica/ModelicaSpecification/pull/2999: It is currently not possible for a tool to know for sure when a model is intended to be a simulation model.  This means that it is not until the user requests an operation such as simulating the model that the tool can know for sure that the model is supposed to be treated as a simulation model.  Drawbacks of this include:
- If the user asks the tool to "validate" all models in a package, the tool doesn't know which of these that should meet the higher standards of a simulation model.
- The tool cannot automatically compile a list of all simulation models in a library.

Having factored out https://github.com/modelica/ModelicaSpecification/issues/3435 as a separate issue, the suggested solution from the discussion in https://github.com/modelica/ModelicaSpecification/pull/2999 for turning _simulation model_ into something that tools can check would be to define it as a class that:
- is a model or block
- is not partial
- has `StopTime`

Considering the amount of discussion we've had around this topic, the specification should clarify that it is not forbidden to simulate models even if they don't fulfill the requirements of a _simulation model_.

**Comment by HansOlsson** [comment]
> Breaking out this concern from #2999: It is currently not possible for a tool to know for sure when a model is intended to be a simulation model. This means that it is not until the user requests an operation such as simulating the model that the tool can know for sure that the model is supposed to be treated as a simulation model. Drawbacks of this include:
> 
> * If the user asks the tool to "validate" all models in a package, the tool doesn't know which of these that should meet the higher standards of a simulation model.
> * The tool cannot automatically compile a list of all simulation models in a library.
> 
> Having factored out #3435 as a separate issue, the suggested solution from the discussion in #2999 for turning _simulation model_ into something that tools can check would be to define it as a class that:
> 
> * is a model or block
> * is not partial
> * has `StopTime`
> 
> Considering the amount of discussion we've had around this topic, the specification should clarify that it is not forbidden to simulate models even if they don't fulfill the requirements of a _simulation model_.

That doesn't correspond to the current use of _simulation model_, but to something else (see the end).

Instead _simulation model_ is currently used to define a list of requirement for a model in order that it can be simulated, such as used constants having a value, discrete variables being assigned in when-clauses, outer/inner requirement, virtual connection graph having proper roots etc. It doesn't make sense to simulate models that don't satisfy those requirements.

Thus the proper alternative (that still needs to be analyzed) is that if the model or block has `StopTime` and is not partial, _it is required to be a simulation model_. That implies that tools can still check (including simulating) the same models as proposed - but also that the current uses of _simulation model_ remain valid.

With that definition there might be additional models that satisfy the requirement of a simulation model, and even can be simulated. E.g., Modelica.Blocks.Sources.Step However, they are not that interesting, and it doesn't make sense to e.g., test and compute coverage based on them.

We could introduce a different term than _simulation model_ to indicate a model or block that has `StopTime` and is not partial; such as "model proposed to be simulated" and say that "a model proposed to be simulated must be a simulation model, but to simulate a model it is only necessary that it satisfies the requirements for a simulation model"; but to me that it is getting too verbose.

**Comment by HansOlsson** [comment]
Thinking more we could, of course, also introduce another name for the current "simulation model" - and instead call the new concept "simulation model"; but to me that seems likely to lead to more confusion.

As I see we can then either close this one without change, or introduce a term for "model proposed to be simulated" as above.

**Comment by henrikt-ma** [comment]
I kind of like the idea of introducing another name for the current "simulation model", but in order to even start thinking of what could be a suitable name I'd need to first understand what it is that this concept captures.  Is it that the model is syntactically correct, globally balanced, and free of structural singularities?  Then we could perhaps call them _translatable_ models.

What are the situations where this concept would be used?  If we add the rule _a simulation model shall be translatable_, what else is there to say?

**Comment by HansOlsson** [comment]
> I kind of like the idea of introducing another name for the current "simulation model", but in order to even start thinking of what could be a suitable name I'd need to first understand what it is that this concept captures. Is it that the model is syntactically correct, globally balanced, and free of structural singularities? Then we could perhaps call them _translatable_ models.

I assume that is in contrast to _simulateable_, but even if I could see some differences between translateable and simulateable, I don't think we need both of them - and I think the important one is _simulateable_ - or basically if you cannot simulate the model it doesn't make sense to translate it.

Consider the following cases:
- `assert(false, "Wrong choice");` (assumedly the assert was originally conditional in some way). Technically, one could argue that it can be translated, even if not simulated. Users appreciate getting early diagnostics of this; not waiting for the entire translation.
- The same with a complicated condition (that is always false), so in one sense it depends on what tools evaluate during translation.
- Hidden structural singularities. Since they are hidden it may not be detected during translation, but it will still cause problems during simulation. If a tool figures out how to detect them and stop the translation I view that as a good thing, not as a failure.

Thus I view the simulation model requirements above (globally balanced, free of structural singularities, roots for virtual connection graphs) as minimum requirements that are easy to check. 

So simulation model mean that model is intended to simulate and the minimum requirements as easy checks, and I don't see a need for a separate term for models that are only intended to be translated. (Even FMUs are intended to be part of simulations.)

Thus I propose to close this without changes.
